### PR TITLE
[ADD] membership_prorrate_variable_period

### PR DESCRIPTION
**Pull request by pedrobaeza**
# 
# Prorrate the membership fee with variable periods

Helper module for making the prorrate correctly when you are dealing with
variable periods in your memberships, which functionality is provided by
the module _membership_variable_period_.
# Installation

This module is auto-installed if you have _membership_variable_period_ and
_membership_prorrate_ modules installed.
# Usage

For leading with variable periods and prorrate, only weekly, monthly or yearly
periods are allowed. When you create an invoice with a membership product
with variable period, the theoretical period is calculated from the invoice
date (or today if empty) as the natural period of the selected unit that
includes that date. For example, if the invoice is on the 15th of January
and the unit is month, the theoretical period to invoice will be all the month
of January.
# Known issues / Roadmap
- Allow to handle periods of months that make the year divisible: 2 months
  (two-monthly), 3 months (quartely), 4 months (four-monthly) or 6 months
  (biannual).
- On weekly memberships, consider the duality sunday/monday start of the week.

### CODE REVIEW

**Comment by rafaelbn** [comment]
In runbot when installing module sale for testing sales orders and contract we got this error:

```
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/http.py", line 574, in dispatch
    result = self._call_function(**self.params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/service/model.py", line 113, in wrapper
    return f(dbname, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/http.py", line 307, in checked_call
    return self.endpoint(*a, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/http.py", line 803, in __call__
    return self.method(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/http.py", line 403, in response_wrap
    response = f(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/addons/web/controllers/main.py", line 948, in call_button
    action = self._call_kw(model, method, args, {})
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/addons/base/module/module.py", line 446, in button_immediate_install
    return self._button_immediate_function(cr, uid, ids, self.button_install, context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/addons/base/module/module.py", line 494, in _button_immediate_function
    registry = openerp.modules.registry.RegistryManager.new(cr.dbname, update_module=True)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/modules/registry.py", line 370, in new
    openerp.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/modules/loading.py", line 355, in load_modules
    loaded_modules, update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/modules/loading.py", line 255, in load_marked_modules
    loaded, processed = load_module_graph(cr, graph, progressdict, report=report, skip_modules=loaded_modules, perform_checks=perform_checks)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/modules/loading.py", line 179, in load_module_graph
    _load_data(cr, module_name, idref, mode, kind='demo')
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/modules/loading.py", line 118, in _load_data
    tools.convert_file(cr, module_name, filename, idref, mode, noupdate, kind, report)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/tools/convert.py", line 901, in convert_file
    convert_xml_import(cr, module, fp, idref, mode, noupdate, report)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/tools/convert.py", line 987, in convert_xml_import
    obj.parse(doc.getroot(), mode=mode)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/tools/convert.py", line 853, in parse
    self._tags[rec.tag](self.cr, rec, n, mode=mode)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/tools/convert.py", line 763, in _tag_record
    id = self.pool['ir.model.data']._update(cr, self.uid, rec_model, self.module, res, rec_id or False, not self.isnoupdate(data_node), noupdate=self.isnoupdate(data_node), mode=self.mode, context=rec_context )
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/addons/base/ir/ir_model.py", line 1077, in _update
    res_id = model_obj.create(cr, uid, values, context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/api.py", line 336, in old_api
    result = method(recs, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/addons/membership_variable_period/models/account_invoice.py", line 77, in create
    line = super(AccountInvoiceLine, self).create(vals)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/addons/membership_prorrate/models/account_invoice.py", line 42, in create
    invoice_line = super(AccountInvoiceLine, self).create(vals)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/addons/membership_initial_fee/models/account_invoice.py", line 37, in create
    product = self.env['product.product'].browse(vals['product_id'])
ParseError: "product_id" while parsing /srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114543-4-c4d71b/openerp/addons/sale/sale_demo.xml:372, near
<record id="test_crm_sale_invoice_1_line_1" model="account.invoice.line">
            <field name="name">Basic computer with Dvorak keyboard and left-handed mouse</field>
            <field name="invoice_id" ref="test_crm_sale_invoice_1"/>
            <field name="price_unit">250</field>
            <field name="quantity">1</field>
            <field name="account_id" ref="account.a_sale"/>
        </record>
```

**Comment by pedrobaeza** [comment]
@rafaelbn, that was probably because the branch needed a rebase. Try now again.

**Comment by pedrobaeza** [comment]
Any news on this?

**Comment by rafaelbn** [comment]
@janetrp please test this one. Rebuilding for testing!

**Comment by janetrp** [comment]
same error in the installation of Sales Module

**Comment by rafaelbn** [comment]
```
2015-11-03 17:14:37 	ERROR 	server 	openerp.http:540 _handle_exception

Exception during JSON request handling.
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/http.py", line 574, in dispatch
    result = self._call_function(**self.params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/service/model.py", line 118, in wrapper
    return f(dbname, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/http.py", line 307, in checked_call
    return self.endpoint(*a, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/http.py", line 803, in __call__
    return self.method(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/http.py", line 403, in response_wrap
    response = f(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/addons/web/controllers/main.py", line 948, in call_button
    action = self._call_kw(model, method, args, {})
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/addons/base/module/module.py", line 447, in button_immediate_install
    return self._button_immediate_function(cr, uid, ids, self.button_install, context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/addons/base/module/module.py", line 495, in _button_immediate_function
    registry = openerp.modules.registry.RegistryManager.new(cr.dbname, update_module=True)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/modules/registry.py", line 370, in new
    openerp.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/modules/loading.py", line 354, in load_modules
    loaded_modules, update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/modules/loading.py", line 255, in load_marked_modules
    loaded, processed = load_module_graph(cr, graph, progressdict, report=report, skip_modules=loaded_modules, perform_checks=perform_checks)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/modules/loading.py", line 179, in load_module_graph
    _load_data(cr, module_name, idref, mode, kind='demo')
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/modules/loading.py", line 118, in _load_data
    tools.convert_file(cr, module_name, filename, idref, mode, noupdate, kind, report)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/tools/convert.py", line 901, in convert_file
    convert_xml_import(cr, module, fp, idref, mode, noupdate, report)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/tools/convert.py", line 987, in convert_xml_import
    obj.parse(doc.getroot(), mode=mode)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/tools/convert.py", line 853, in parse
    self._tags[rec.tag](self.cr, rec, n, mode=mode)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/tools/convert.py", line 763, in _tag_record
    id = self.pool['ir.model.data']._update(cr, self.uid, rec_model, self.module, res, rec_id or False, not self.isnoupdate(data_node), noupdate=self.isnoupdate(data_node), mode=self.mode, context=rec_context )
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/addons/base/ir/ir_model.py", line 1052, in _update
    res_id = model_obj.create(cr, uid, values, context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/api.py", line 250, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/api.py", line 345, in old_api
    result = method(recs, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/addons/membership_variable_period/models/account_invoice.py", line 77, in create
    line = super(AccountInvoiceLine, self).create(vals)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/api.py", line 248, in wrapper
    return new_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/addons/membership_prorrate/models/account_invoice.py", line 43, in create
    invoice_line = super(AccountInvoiceLine, self).create(vals)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/api.py", line 248, in wrapper
    return new_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/addons/membership_initial_fee/models/account_invoice.py", line 37, in create
    product = self.env['product.product'].browse(vals['product_id'])
ParseError: "product_id" while parsing /srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3118847-4-763fec/openerp/addons/sale/sale_demo.xml:372, near
<record id="test_crm_sale_invoice_1_line_1" model="account.invoice.line">
            <field name="name">Basic computer with Dvorak keyboard and left-handed mouse</field>
            <field name="invoice_id" ref="test_crm_sale_invoice_1"/>
            <field name="price_unit">250</field>
            <field name="quantity">1</field>
            <field name="account_id" ref="account.a_sale"/>
```

**Comment by pedrobaeza** [comment]
OK, error fixed. Please try now.

**Comment by janetrp** [comment]
Tested in Runbot
Association end date does not appear.
2015-11-04 17_50_03-odoo

**Comment by rafaelbn** [comment]
Hi @pedrobaeza any news one this? @janetrp could you review again please? Thanks

**Comment by janetrp** [comment]
runbot is rebuilding

File: membership_initial_fee/models/account_invoice.py [hunk 1]
```diff
@@ -34,7 +34,8 @@ def _prepare_initial_fee_vals(self, invoice_line):
     @api.model
     def create(self, vals):
         invoice_line = super(AccountInvoiceLine, self).create(vals)
-        product = self.env['product.product'].browse(vals['product_id'])
+        product = self.env['product.product'].browse(
+            vals.get('product_id'))
```
Comment (antespi) [inline membership_initial_fee/models/account_invoice.py:38]: Why don't you use this?

```
product = self.env['product.product'].browse(vals.get('product_id', []))
```

like you do in `membership_prorrate/models/account_invoice.py:44`

File: membership_prorrate_variable_period/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# (c) 2015 Antiun Ingeniería S.L. - Pedro M. Baeza
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    'name': 'Prorrate membership fee for variable periods',
+    'version': '1.0',
```
Comment (antespi) [inline membership_prorrate_variable_period/__openerp__.py:7]: 8.0.1.0.0

**Comment by antespi** [comment]
After comments resolved, :+1: tested in runbot

**Comment by janetrp** [comment]
error remains, end date of membership is not appearing and membership is not aplying.
2015-12-03 13_39_45-customers - odoo

**Comment by rafaelbn** [comment]
Hi @pedrobaeza could you review this please? Thanks

By the way, there is not standard labels and milestone in this repo

**Comment by pedrobaeza** [comment]
This should be fixed now

**Comment by JMODaERP** [comment]
I tested in runbot :+1:

**Comment by JavierIniesta** [comment]
Tested in runbot :+1:

**Comment by rafaelbn** [comment]
grear thanks :+1:

**Pull request merged**